### PR TITLE
(in progress): sol["cost"] should have units

### PR DESCRIPTION
**Pull request by whoburg**
WIP resolving #800. So far this just adds an aspirational test. Work on giving SolutionArray["cost"] units should be done on this PR branch.

### CODE REVIEW

**Comment by bqpd** [comment]
@whoburg, ready for review

File: gpkit/small_classes.py [hunk 1]
```diff
@@ -134,10 +134,9 @@ def _enray_and_unit_dict(d_in, d_out, unitless_keys=(), united=False):
     "Recursively turns lists into numpy arrays."
     for k, v in d_in.items():
         if isinstance(v, dict):
-            if k in unitless_keys:
-                united = False
+            subunited = united and k not in unitless_keys
             d_out[k] = _enray_and_unit_dict(v, v.__class__(),
-                                            unitless_keys, united)
+                                            unitless_keys, subunited)
```
Comment (whoburg) [inline gpkit/small_classes.py]: phew, I've been wondering about this for awhile.

File: gpkit/solution_array.py [hunk 2]
```diff
@@ -112,12 +112,12 @@ def table(self, tables=("cost", "sweepvariables", "freevariables",
...
                 strs += ["\n%s\n----" % table_title]
                 if len(self) > 1:
-                    costs = ["%-8.3g" % c for c in subdict[:4]]
+                    costs = ["%-8.3g" % c for c in mag(subdict[:4])]
                     strs += [" [ %s %s ]" % ("  ".join(costs),
                                              "..." if len(self) > 4 else "")]
                     cost_units = self.program[0].cost.units
                 else:
-                    strs += [" %-.4g" % subdict]
+                    strs += [" %-.4g" % mag(subdict)]
```
Comment (whoburg) [inline gpkit/solution_array.py]: It seems overly complicated that we take the `mag()` of a Quantity here and then tack the `unitstr` back on below at line ~127. Can we use pint's pretty string formatting (http://pint.readthedocs.io/en/0.6/tutorial.html#string-formatting) to clean this up?

File: gpkit/tests/t_model.py [hunk 3]
```diff
@@ -76,20 +76,30 @@ def test_cost_freeing(self):
...
+        t = Variable("t", 10, "hr")
+        T_loiter = Variable("T_{loiter}", 1, "hr")
+        T_reserve = Variable("T_{reserve}", 45, "min")
+        M = VectorVariable(2, "M")
+
+        prob = Model(1/R,
+                     [t >= sum(R/a0/M/theta**0.5) + T_loiter + T_reserve,
+                      M <= 0.76])
+        sol = prob.solve(solver=self.solver, verbosity=0)
+        # TODO: add better way to test united things
```
Comment (whoburg) [inline gpkit/tests/t_model.py]: see https://github.com/hgrecco/pint/issues/347

File: gpkit/tests/t_model.py [hunk 4]
```diff
@@ -76,20 +76,30 @@ def test_cost_freeing(self):
...
+        costdiff = (sol["cost"] - 0.0005532/R.units).magnitude
+        self.assertAlmostEqual(costdiff, 0, self.ndig)
+        diff = (sol["constants"]["a0"] - 340.29*a0.units).magnitude
+        self.assertAlmostEqual(diff, 0, self.ndig)
+        diff = (sol["variables"]["a0"] - 340.29*a0.units).magnitude
+        self.assertAlmostEqual(diff, 0, self.ndig)
+        diff = (sol["freevariables"]["R"] - 1807.58*R.units).magnitude
+        # inverse of cost is much more sensitive than self.ndig, unsurprisingly
```
Comment (whoburg) [inline gpkit/tests/t_model.py]: on all these `assertAlmostEqual`'s, how about comparing the ratio to one? Something like

```python
self.assertAlmostEqual((sol["constants"]["a0"] / (340.29*a0.units)).to("dimensionless"), 1.0)
```

File: gpkit/tests/t_sub.py [hunk 5]
```diff
@@ -233,16 +233,19 @@ def __init__(self):
...
         if not isinstance(a["x"].key.units, str):
-            self.assertAlmostEqual(a.solve(verbosity=0)["cost"], 0.3333333)
-            self.assertAlmostEqual(b.solve(verbosity=0)["cost"], 0.01)
-            self.assertAlmostEqual(concat_cost, 0.0109361)  # 1 cm/1 yd
+            costdiffa = mag(1/gpkit.units.yd - a.solve(verbosity=0)["cost"])
+            self.assertAlmostEqual(costdiffa, 0, 5)
+            costdiffb = mag(1*gpkit.units.cm - b.solve(verbosity=0)["cost"])
+            self.assertAlmostEqual(costdiffb, 0, 5)
+            costdiffc = mag(1*gpkit.units.cm/gpkit.units.yd - concat_cost)
+            self.assertAlmostEqual(costdiffc, 0, 5)
```
Comment (whoburg) [inline gpkit/tests/t_sub.py]: same comment here, compare ratios to 1.0?

**Comment by whoburg** [comment]
code review complete, back to @bqpd

File: gpkit/solution_array.py (hunk 2 above)
Comment (bqpd) [inline gpkit/solution_array.py]: I haven't yet gotten pint's units to look the way I want, and this lets us be consistent with nomial units printing.

File: gpkit/tests/t_model.py (hunk 4 above)
Comment (bqpd) [inline gpkit/tests/t_model.py]: Works for me!

File: gpkit/tests/t_model.py (hunk 3 above)
Comment (bqpd) [inline gpkit/tests/t_model.py]: Thought about using that, but preferred using the relative tolerance as below.

**Comment by bqpd** [comment]
@whoburg, ready for rereview

**Comment by galbramc** [comment]
test this please (just checking the windozse failure)

**Pull request merged**